streaming/base/shared/prefix.py

```diff
@@ -54,48 +62,87 @@ def _unpack_locals(data: bytes) -> Set[str]:
     return set(text.split('\0'))
 
 
-def get_shm_prefix(my_locals: List[str], world: World) -> Tuple[str, SharedMemory]:
-    """Register or lookup our shared memory prefix.
+def _check_self(my_locals: List[str]) -> Set[str]:
+    """Check our local working directories for overlap.
 
     Args:
-        my_locals (List[str]): Local working dir of each stream, relative to /. We need to verify
-            that there is no overlap with any other currently running StreamingDataset.
-        world (World): Information about nodes, ranks, and workers.
+        my_locals (List[str]): Local dirs.
 
     Returns:
-        Tuple[str, CreateSharedMemory]: Shared memory prefix and object. The name is required to be
-            very short for Mac OSX.
+        Set[str]: Local dirs as a set.
     """
-    # Check my locals for overlap.
     my_locals_set = set()
     for dirname in my_locals:
         if dirname in my_locals_set:
             raise ValueError(f'Reused local directory: {dirname}. Provide a different one.')
         my_locals_set.add(dirname)
+    return my_locals_set
 
-    # Local leader goes first, checking and registering.
-    if world.is_local_leader:
-        # Local leader walks the existing shm prefixes starting from zero, verifying that there is
-        # no local working directory overlap.  When attaching to an existing shm fails, we have
-        # reached the end of the existing shms.
-        for prefix_int in range(10**6):
-            prefix = f'{prefix_int:06}'
-            name = f'{prefix}_locals'
-            try:
-                shm = CreateSharedMemory(name, False).shm
-            except:
-                break
-            their_locals_set = _unpack_locals(bytes(shm.buf))
-            both = my_locals_set & their_locals_set
-            if both:
-                raise ValueError(f'Reused local directory: {sorted(my_locals_set)} vs ' +
-                                 f'{sorted(their_locals_set)}. Provide a different one.')
 
-        # Local leader registers the first available shm prefix, recording its locals.
+def _check_and_find(my_locals_set: Set[str]) -> int:
+    """Find the next available prefix while checking existing dirs for overlap.
+
+    Local leader walks the existing shm prefixes starting from zero, verifying that there is no
+    local working directory overlap. When attaching to an existing shm fails, we have reached the
+    end of the existing shms. We will register the next one.
+    """
+    prefix_int = 0
+    for prefix_int in _each_prefix_int():
+        prefix = f'{prefix_int:06}'
+        name = f'{prefix}_locals'
+        try:
+            shm = SharedMemory(name, False)
+        except:
```

Comment:
any specific exception you want to catch such as `FileNotFoundError` or `FileExistError` ?

Reply:
done